distributed/tests/test_active_memory_manager.py

```diff
@@ -796,3 +1009,48 @@ async def test_ReduceReplicas_stress(c, s, *nannies):
     policy must not disrupt the computation too much.
     """
     await tensordot_stress(c)
+
+
+# @pytest.mark.slow
+@pytest.mark.avoid_ci(reason="distributed#5371")
+@pytest.mark.parametrize("use_ReduceReplicas", [False, True])
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)] * 10,
+    Worker=Nanny,
+    config={
+        "distributed.scheduler.active-memory-manager.start": True,
+        # If interval is too low, then the AMM will rerun while tasks have not yet have
+        # the time to migrate. This is OK if it happens occasionally, but if this
+        # setting is too aggressive the cluster will get flooded with repeated comm
+        # requests.
```

Comment:
Is this a problem in real-world clusters too?

Reply:
Yes. As the comment says, it's OK if once in a while a task takes more than 2 seconds to migrate; it *could* be problematic if many tasks over and over take longer than 2 seconds. I don't really have a pulse about how long it typically takes on a big busy cluster between get_data and the actual data arriving.

Reply:
On a busy cluster, it wouldn't surprise me if the full roundtrip (including the scheduler finding out the task had migrated) could take longer that 2s, especially when data is spilled to disk, and especially if multiple workers are retiring (scale-down).

I suppose we'll just have to try it and find out. I could see it becoming necessary to have some memory on the AMM and not drop `self.pending` every cycle (or store `pending` as `prev` for the next cycle, and remove any current suggestions that were in `prev` or something), but that will be a future PR.

Reply:
Can we see this when it happens? Are there logs or other traces or would we be blind if that happened?

Reply:
To clarify: this is exclusively an issue if
- a substantial amount of keys are not replicated within 2 seconds, AND
- something *other than the AMM*, such as a computation or a new worker coming online, tilted the memory balance of the cluster enough that the AMM will take a substantially different decision from the previous one

In all other cases, including when some/most of the keys that the AMM asked to be replicated have already moved and increased the memory usage in the recipient, then the AMM will just take the same decision as before.

Again, it is OK if this happens occasionally - you'll just end up with a bunch of superfluous replications which will be reverted shortly by ReduceReplicas.

> Can we see this when it happens? 

At every AMM iteration, in the debug log you can read how many keys each worker is going to receive and drop. You may witness a worker that was previously unaffected suddenly loaded with replication requests.

e.g.
FIRST ITERATION:
```
distributed.active_memory_manager - DEBUG - Running policy: ReduceReplicas()
distributed.active_memory_manager - DEBUG - Running policy: RetireWorker('tcp://127.0.0.1:33379')
distributed.active_memory_manager - DEBUG - Enacting suggestions for 143 tasks:
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34809', name: tcp://127.0.0.1:34809, status: running, memory: 1, processing: 0> to acquire 95 replicas
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34810', name: tcp://127.0.0.1:34810, status: running, memory: 1, processing: 0> to acquire 48 replicas
distributed.active_memory_manager - DEBUG - Active Memory Manager run in 1ms
```

HEALTHY second iteration: most of the tasks have been replicated within 2s
```
distributed.active_memory_manager - DEBUG - Running policy: ReduceReplicas()
distributed.active_memory_manager - DEBUG - Running policy: RetireWorker('tcp://127.0.0.1:33379')
distributed.active_memory_manager - DEBUG - Enacting suggestions for 12 tasks:
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34809', name: tcp://127.0.0.1:34809, status: running, memory: 1, processing: 0> to acquire 12 replicas
distributed.active_memory_manager - DEBUG - Active Memory Manager run in 1ms
```

HEALTHY second iteration: network and/or the workers' event loop are slow, but nothing besides the AMM itself substantially changed the cluster balance
```
distributed.active_memory_manager - DEBUG - Running policy: ReduceReplicas()
distributed.active_memory_manager - DEBUG - Running policy: RetireWorker('tcp://127.0.0.1:33379')
distributed.active_memory_manager - DEBUG - Enacting suggestions for 135 tasks:
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34809', name: tcp://127.0.0.1:34809, status: running, memory: 1, processing: 0> to acquire 92 replicas
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34810', name: tcp://127.0.0.1:34810, status: running, memory: 1, processing: 0> to acquire 43 replicas
distributed.active_memory_manager - DEBUG - Active Memory Manager run in 1ms
```

UNHEALTHY second iteration: network and/or the workers' event loop are slow, and a computation suddenly saturated the first recipient, so the AMM now decides to load the second recipient a lot more. When (if) eventually the replicas on the first recipient finally go through, we'll end up with a bunch of duplicates which will be removed by ReduceReplicas.
```
distributed.active_memory_manager - DEBUG - Running policy: ReduceReplicas()
distributed.active_memory_manager - DEBUG - Running policy: RetireWorker('tcp://127.0.0.1:33379')
distributed.active_memory_manager - DEBUG - Enacting suggestions for 135 tasks:
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34809', name: tcp://127.0.0.1:34809, status: running, memory: 1, processing: 0> to acquire 20 replicas
distributed.active_memory_manager - DEBUG - - <WorkerState 'tcp://127.0.0.1:34810', name: tcp://127.0.0.1:34810, status: running, memory: 1, processing: 0> to acquire 115 replicas
distributed.active_memory_manager - DEBUG - Active Memory Manager run in 1ms
```

Reply:
The only thing I'm worried about is that 2s is very small for key replication imo. network latencies, network bandwidth, (de-)serialization and diskIO all play together here. Especially if a single worker is required to offload a few GB of data that will very likely take more than 2s, wouldn't it?

Debug logs are pretty hidden but I'm not familiar enough with the system, yet, to make a better suggestion. I was wondering if there is a condition that would allow us to detect such a situation and then log a meaningful warning.

Reply:
Either way, we can defer this to another PR/issue. I'm mostly wondering if this is something we should follow up on

Reply:
I could keep track of incomplete replications from previous iterations and don't reiterate them for e.g. 30 seconds